File: 2471-minimum-number-of-operations-to-sort-a-binary-tree-by-level/2471-minimum-number-of-operations-to-sort-a-binary-tree-by-level.py
from collections import deque

class Solution:
    def minimumOperations(self, root):
        q = deque([root])
        swp = 0
        
        while q:
            n = len(q)
            v = []
            
            for _ in range(n):
                cur = q.popleft()
                v.append(cur.val)
                if cur.left:
                    q.append(cur.left)
                if cur.right:
                    q.append(cur.right)
            
            sorted_v = sorted(v)
            idx_map = {val: i for i, val in enumerate(v)}
            
            for i in range(n):
                if v[i] != sorted_v[i]:
                    idx = idx_map[sorted_v[i]]
                    idx_map[v[i]] = idx
                    v[i], v[idx] = v[idx], v[i]
                    swp += 1
        
        return swp

File: 2471-minimum-number-of-operations-to-sort-a-binary-tree-by-level/2471-minimum-number-of-operations-to-sort-a-binary-tree-by-level.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minimumOperations(TreeNode* root) {
        queue<TreeNode*>q;
        int swp = 0;
        q.push(root);
        while(q.size()){
            int n = q.size();
            vector<int>v(n);
            for(int i = 0 ; i < n;i++){
                auto cur = q.front();q.pop();
                v[i] = (cur->val);
                if(cur->left)q.push(cur->left);
                if(cur->right)q.push(cur->right);
            }
            vector<int>sorted=v;
            sort(sorted.begin(),sorted.end());
            unordered_map<int,int>mp;
            for(int i = 0 ; i < n;i++)mp[v[i]] = i;

            for(int i = 0 ; i < n;i++){
                if(v[i]!=sorted[i]){
                    int idx = mp[sorted[i]];
                    mp[v[i]] = idx;
                    swap(v[i],v[idx]);
                    swp++;
                }
            }
        }
        return swp;

    }
};